flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClientFactory.java

```diff
@@ -81,16 +92,12 @@ NettyPartitionRequestClient createPartitionRequestClient(ConnectionID connection
 				Object old = clients.putIfAbsent(connectionId, connectingChannel);
 
 				if (old == null) {
-					nettyClient.connect(connectionId.getAddress()).addListener(connectingChannel);
-
-					client = connectingChannel.waitForChannel();
-
-					clients.replace(connectionId, connectingChannel, client);
+					connectChannelWithRetry(connectingChannel, connectionId);
+					client = (NettyPartitionRequestClient) clients.get(connectionId);
 				}
 				else if (old instanceof ConnectingChannel) {
```

Comment:
we can merge `if (old == null || old instanceof ConnectionChannel)` because the actions are actually the same.